compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/desugar/Desugar.java

```diff
@@ -4994,15 +4982,43 @@ private void analyzeOnFailClause(BLangOnFailClause onFailClause, BLangBlockStmt
             } else {
                 rewrite(onFailClause, env);
             }
-            if (markClosures || !onFailClause.bodyContainsFail) {
-                onFailClause.possibleClosureSymbols.forEach(symbol -> symbol.closure = true);
+        }
+    }
+
+    private BLangFail createOnFailInvocation(BLangOnFailClause onFailClause, BLangFail fail) {
+        BLangBlockStmt onFailBody = ASTBuilderUtil.createBlockStmt(onFailClause.pos);
+        onFailBody.stmts.addAll(onFailClause.body.stmts);
+        env.scope.entries.putAll(onFailClause.body.scope.entries);
+        onFailBody.failureBreakMode = onFailClause.body.failureBreakMode;
+
+        BVarSymbol onFailErrorVariableSymbol =
+                ((BLangSimpleVariableDef) onFailClause.variableDefinitionNode).var.symbol;
+        BLangSimpleVariable errorVar = ASTBuilderUtil.createVariable(onFailErrorVariableSymbol.pos,
+                onFailErrorVariableSymbol.name.value, onFailErrorVariableSymbol.type, rewrite(fail.expr, env),
+                onFailErrorVariableSymbol);
+        BLangSimpleVariableDef errorVarDef = ASTBuilderUtil.createVariableDef(onFailClause.pos, errorVar);
+        env.scope.define(onFailErrorVariableSymbol.name, onFailErrorVariableSymbol);
+        onFailBody.stmts.add(0, errorVarDef);
+
+        if (onFailClause.isInternal && fail.exprStmt != null) {
```

Comment:
This flow comes under transaction statements. Hence tests are already in the transaction internal repo